### ISSUE TITLE
Random Lua Errors - "Wear" stops Working

### ISSUE THREAD

**Issue by Hazbelll** [CapsAdmin/pac3#200]
```
[ERROR] lua/pac3/core/client/base_part.lua:76: table index is nil
  1. SetUniqueID - lua/pac3/core/client/base_part.lua:76
   2. SetTable - lua/pac3/core/client/base_part.lua:619
    3. unknown - lua/pac3/editor/client/wear.lua:100

Timer Failed! [Simple][@lua/pac3/editor/client/wear.lua (line 95)]

[ERROR] lua/includes/extensions/net.lua:180: net.ReadType: Couldn't read type 128
  1. error - [C]:-1
   2. ReadType - lua/includes/extensions/net.lua:180
    3. ReadType - lua/includes/extensions/net.lua:121
     4. ReadType - lua/includes/extensions/net.lua:124
      5. ReadTable - lua/includes/extensions/net.lua:124
       6. func - lua/pac3/editor/client/wear.lua:169
        7. unknown - lua/includes/extensions/net.lua:32


[ERROR] lua/includes/extensions/net.lua:180: net.ReadType: Couldn't read type 235
  1. error - [C]:-1
   2. ReadType - lua/includes/extensions/net.lua:180
    3. ReadType - lua/includes/extensions/net.lua:121
     4. ReadType - lua/includes/extensions/net.lua:124
      5. ReadTable - lua/includes/extensions/net.lua:124
       6. func - lua/pac3/editor/client/wear.lua:169
        7. unknown - lua/includes/extensions/net.lua:32


[ERROR] lua/includes/extensions/net.lua:180: net.ReadType: Couldn't read type 96
  1. error - [C]:-1
   2. ReadType - lua/includes/extensions/net.lua:180
    3. ReadTable - lua/includes/extensions/net.lua:121
     4. func - lua/pac3/editor/client/wear.lua:169
      5. unknown - lua/includes/extensions/net.lua:32


[ERROR] lua/includes/extensions/net.lua:180: net.ReadType: Couldn't read type 2
  1. error - [C]:-1
   2. ReadType - lua/includes/extensions/net.lua:180
    3. ReadType - lua/includes/extensions/net.lua:121
     4. ReadType - lua/includes/extensions/net.lua:124
      5. ReadTable - lua/includes/extensions/net.lua:124
       6. func - lua/pac3/editor/client/wear.lua:169
        7. unknown - lua/includes/extensions/net.lua:32

tried to create unknown function: 0x19b84a78 "grou�8"!
pac3 tried to create unknown part grou�8

[ERROR] lua/pac3/core/client/base_part.lua:76: table index is nil
  1. SetUniqueID - lua/pac3/core/client/base_part.lua:76
   2. SetTable - lua/pac3/core/client/base_part.lua:619
    3. unknown - lua/pac3/editor/client/wear.lua:100

Timer Failed! [Simple][@lua/pac3/editor/client/wear.lua (line 95)]


[ERROR] lua/pac3/core/client/base_part.lua:751: table index is nil
  1. Remove - lua/pac3/core/client/base_part.lua:751
   2. unknown - lua/pac3/core/client/hooks.lua:394

Timer Failed! [pac_gc][@lua/pac3/core/client/hooks.lua (line 391)]

[PAC3] OGG error: stream[0x42a359c8][40][https://copy.com/BLNC1joOrF3lYpUR] reason: download failed

[ERROR] lua/includes/extensions/net.lua:180: net.ReadType: Couldn't read type 64
  1. error - [C]:-1
   2. ReadType - lua/includes/extensions/net.lua:180
    3. ReadType - lua/includes/extensions/net.lua:121
     4. ReadType - lua/includes/extensions/net.lua:124
      5. ReadType - lua/includes/extensions/net.lua:124
       6. ReadTable - lua/includes/extensions/net.lua:124
        7. func - lua/pac3/editor/client/wear.lua:169
         8. unknown - lua/includes/extensions/net.lua:32

[ERROR] lua/pac3/editor/client/wear.lua:86: attempt to index field 'self' (a nil value)
  1. WearPartFromServer - lua/pac3/editor/client/wear.lua:86
   2. go - lua/pac3/editor/client/wear.lua:133
    3. unknown - lua/pac3/editor/client/wear.lua:150

Timer Failed! [pac_wear_queue][@lua/pac3/editor/client/wear.lua (line 141)]


[ERROR] lua/includes/extensions/net.lua:180: net.ReadType: Couldn't read type 217
  1. error - [C]:-1
   2. ReadType - lua/includes/extensions/net.lua:180
    3. ReadType - lua/includes/extensions/net.lua:124
     4. ReadType - lua/includes/extensions/net.lua:124
      5. ReadTable - lua/includes/extensions/net.lua:124
       6. func - lua/pac3/editor/client/wear.lua:169
        7. unknown - lua/includes/extensions/net.lua:32

[ERROR] lua/includes/extensions/net.lua:180: net.ReadType: Couldn't read type 20
  1. error - [C]:-1
   2. ReadType - lua/includes/extensions/net.lua:180
    3. ReadType - lua/includes/extensions/net.lua:124
     4. ReadType - lua/includes/extensions/net.lua:124
      5. ReadTable - lua/includes/extensions/net.lua:124
       6. func - lua/pac3/editor/client/wear.lua:169
        7. unknown - lua/includes/extensions/net.lua:32
```

All of the above got spat out into mine and all other Player's Client Consoles. This occurred randomly on my Server with 30 players on. The Server has been up since 6AM today (GMT +1). The time currently is 06:25 PM (GMT +0).

This has totally broken "Wearing" of PACs on the Server. Players may still use their PACs clientside, but they cannot wear them serverside.

This issue has occurred in the past, totally at random, although rare. I've only acknowledged it happening if `PrePACConfigApply` or `PrePACEditorOpen` are hooked in to. To fix wear not working, the Server must be restarted manually.

My Server uses the Workshop version of PAC with the `wear.lua` in an addons folder to override Workshop's from this commit: https://github.com/CapsAdmin/pac3/commit/fd399dda82e1af81760f7d0ee2dfa2d20951b046

My apologies if this information is vague, but it's all I have.

**Comment by LUModder** [comment]
1. Does the server have an anticheat?
2. Are you able to get the outfit causing this error? (for reproduction
purposes of course)

**Comment by Hazbelll** [comment]
1. No.
2. Possibly. I think I may know who's PAC caused it, but I would have to ask them if they'd be willing to share it with me to pass onto you. I'll look into that.

**Comment by Hazbelll** [comment]
Okay, they kindly provided it:

10:57 AM - Hazbelll: Hey, would you be able to send me that PAC you had on the Server when it bugged out? The PAC devs asked for it to examine and debug.
12:58 PM - Dubsy: yeah, sure no problem
01:00 PM - Dubsy: https://www.dropbox.com/s/hsupkp8gpyqnd9z/Dubsy%20oc.txt?dl=0
01:02 PM - Dubsy: https://www.dropbox.com/s/js6x196wzkhs4xo/Dubsy%27s%20OC%20few%20songs.txt?dl=0
01:03 PM - Dubsy: second one is how i fixed the errors, basacly i just removed the songs and kep only 1 for the speakers and 1 for the crying thing
01:03 PM - Dubsy: kept*

Hope this helps.

**Comment by Hazbelll** [comment]
Any updates?

**Comment by LUModder** [comment]
Tested last night, no errors.

**Comment by Hazbelll** [comment]
Then it's not solely caused by that PAC then (if it all). This is an ongoing issue though, "wear" stops working on Servers. From observation, this is incredibly rare and only occurs when the Server is heavily populated with Players (25+). I believe using either two of the hooks I mentioned in the Original Post may be a potential cause too.

Again, I apologise I have no solid evidence to show, but this issue is spontaneous and gives little to no leads.

**Comment by thegrb93** [comment]
I've seen this error come up in my lua_errors_server a lot too. I run a simple sandbox server so dunno what it could be.

I'm guessing that some pacs are so big they are exceeding the net library 64kb limit.

**Comment by Hazbelll** [comment]
That makes a lot of sense, especially from looking at where the errors are occurring in the Stack Traces.

**Comment by Hazbelll** [comment]
```
[16:47] Player [19][HerpDerpMapleSerp] ERR: Decoding pac3//pony.txt failed: luadata_decode:13: unexpected symbol near '['
[16:47] gamemodes/darkrp/gamemode/modules/fpp/pp/server/core.lua:382: Tried to use a NULL entity!
  1. SteamID - [C]:-1
   2. CanTool - gamemodes/darkrp/gamemode/modules/fpp/pp/server/core.lua:382
    3. CPPICanTool - gamemodes/darkrp/gamemode/modules/fpp/pp/sh_cppi.lua:66
     4. SubmitPart - lua/pac3/editor/server/wear.lua:118
      5. unknown - lua/pac3/editor/server/wear.lua:50
```

EPOE caught this just before "Wear" ceased to work a few minutes ago as of writing this. Perhaps PAC's not playing so nice with checking in FPP whether a Player can touch whichever entity a player's worn PAC is attached to (or vice versa)? I hope it helps.

**Comment by LUModder** [comment]
Now you see why I hate DarkRP so much, FPP is one of the worst prop protections IMO.

**Comment by thegrb93** [comment]
@LUModder It's not FPP's fault that pac tries to use a null entity.

**Comment by HDG-MrDeeJay** [comment]
@LUModder There is literally no problem with FPP. It is one of the *BEST* prop protection scripts available. If PAC3 and FPP are causing errors with one another, perhaps we should look at PAC3 being the culprit, given we already know it to have many flaws that nobody seems willing to do anything about. Unlimited pac3 content for instance is another of these things nobody really wants to fix.

**Comment by thegrb93** [comment]
It is pac. https://github.com/CapsAdmin/pac3/blob/master/lua/pac3/editor/server/wear.lua#L75 data.owner is NULL

**Comment by HDG-MrDeeJay** [comment]
```LUA
if ent:IsValid()then
```

Holy shit was someone drunk when they wrote that?

**Comment by thegrb93** [comment]
There's nothing wrong with it?

**Comment by LUModder** [comment]
God, when did PAC issues become a flamewar




oh right

**Comment by roboderpy** [comment]
```
[DPP] ERROR: Bad Argument #1 to DPP.CanTool (Player Expected, got NULL Entity).
TO USERS: THIS IS A BUG IN lua/pac3/editor/server/wear.lua:75
``` 
How to trigger: wear a pac where you modify world entities and before it wears quickly disconnect. In queue it would try to wear your PAC (god why) when you are already disconnected.

**Comment by LUModder** [comment]
Seems unrelated and probably due to how badly optimized and written your
code is.

**Comment by roboderpy** [comment]
what. PAC is trying to apply part when part owner is disconnected.

**Comment by LUModder** [comment]
That's not what the error says

**Comment by roboderpy** [comment]
Error says that first argument that is passed to CanTool is NULL. That argument is ```data.owner``` in PAC's code

**Comment by LUModder** [comment]
DPP != PAC3

Give me a genuine error from PAC and I might believe your try hard ass for
once

**Comment by roboderpy** [comment]
Old DPP version
```
[PAC3] VLL:629: Argument is not a player!
  1. error - [C]:-1
   2. error - VLL:629
    3. Assert - [VLL: dpp - dpp/sh_init.lua]:73
     4. AssertPlayer - [VLL: dpp - dpp/sh_init.lua]:58
      5. CanTool - [VLL: dpp - dpp/sh_hooks.lua]:236
       6. CPPICanTool - [VLL: dpp - dpp/sh_cppi.lua]:73
        7. SubmitPart - lua/pac3/editor/server/wear.lua:75
         8. unknown - lua/pac3/editor/server/wear.lua:7
```

**Comment by LUModder** [comment]
You're hooking into too much code and from your previous track record of
mistakes I can't believe it to be a PAC error without GENUINE errors with
ONLY PAC being referenced

**Comment by LUModder** [comment]
As seen before the flame war broke out, the outfit being worn had an error in the file, which makes me believe it's bad error handling on either parts of the addons

**Comment by roboderpy** [comment]
Just. Stop. Fucking. Calling. CPPICanTool. Without. Valid. Player. Argument.

**Comment by LUModder** [comment]
Why does it even need to be called?

**Comment by roboderpy** [comment]
PAC calling it when player modify entity through PAC that is not this player itself. On this line https://github.com/CapsAdmin/pac3/blob/master/lua/pac3/editor/server/wear.lua#L75

**Comment by LUModder** [comment]
God this code is horrendous, just check friend status and not use CanTool.

(Pro tip: pull request and not bitch about it)

**Comment by thegrb93** [comment]
Just put 'if not data.owner:IsValid() then return end' at the beginning of the function.

**Comment by roboderpy** [comment]
Thats best solution, but because flex thinks that EVERYONE ELSE's addons is bugged, he wont do it.

**Comment by LUModder** [comment]
What if data is nil? Do it right.

if not data or not IsValid(data.owner) then return end

**Comment by Hazbelll** [comment]
Out of curiosity, what's the point in having an issue tracker for PAC3, if none of the devs seem to give a shit about anything and always fob users off with "fix it yourself" remarks?

**Comment by roboderpy** [comment]
data.owner never can be nil in this case

**Comment by LUModder** [comment]
The people that want PAC fixed don't know how and the devs have moved on
and CBA to work on it, and in my case I don't have an environment to be
able to work on it if I cared.

**Comment by thegrb93** [comment]
You're not helping, and you can't help. Why are you posting?

**Comment by roboderpy** [comment]
I reported bug above and how to reproduce it clearly as **fuck**, but Flex still says that bug was reported wrong and users broke it for themself

**Comment by Shigbeard** [comment]
So now this has been reported for 2 prop protection addons. @LUModder are you telling us that both prop protection addons have shit code? You aren't validating your entity. Scroll up, this same problem occured and you blamed it on FPP, saying you hate DarkRP because FPP was the wrost prop protection addon. Now you are saying @roboderpy is using a shit prop protection script.

Fix your shit Flex. It's an issue with PAC, not with EVERY PROP PROTECTION ADDON MADE THAT ASSUMES ADDONS VALIDATE THEIR OWN ENTITIES.

**Comment by CapsAdmin** [comment]
Just to be clear LUModder is just some person who happens to comment on almost every issue sometimes helpful and sometimes not at all helpful. The only people with commit access is me, Python1320 and Kamshak.

Although no official agreement was made between me and Python1320 I gave him commit access because I thought they (metastruct) wanted to work on PAC3 but I get the feeling that every time I ask about this they are all busy with other things. I guess there's a fear of making a promise and then keeping it while being officially responsible for PAC3.

I have all the freetime in the world to work on PAC3 but I've just not been feeling it. On top of that I've switched to linux (gmod crashes a lot under linux) and I really don't wanna use windows.
